cognite_synthetic_tags/data_stores.py

```diff
@@ -0,0 +1,287 @@
+    c: CogniteClient,
+    id: Optional[List[int]] = None,
+    external_id: Optional[List[str]] = None,
+    ignore_unknown_ids: bool = True,
+    fill_with=np.nan,
+    **kwargs,
+) -> pd.DataFrame:
+    """
+    Wrapper around `CogniteClient.datapoints.retrieve` which fills in any
+    missing columns (i.e. values in external_ids).
+
+    Calling, `CogniteClient.datapoints.retrieve(..., ignore_unknown_ids=True)`
+    results in missing columns, while `ignore_unknown_ids="fill"` or
+    `"fill,dropna"` raise exceptions.
+    """
+    res = c.datapoints.retrieve(
+        id=id,
+        external_id=external_id,
+        ignore_unknown_ids=ignore_unknown_ids,
+        **kwargs,
+    )
+    df = res.to_pandas()
+
+    # Compile a list of expected columns:
+    tags = external_id or id or []
+    aggregates = kwargs.get("aggregates")
+    columns = (
+        tags
+        if aggregates is None
+        else [f"{tag}|{aggregate}" for tag in tags for aggregate in aggregates]
+    )
+
+    # Add any missing columns:
+    df.loc[:, [c for c in columns if c not in df.columns]] = fill_with
+
+    # Reorder columns:
+    df = df[columns]
+
+    return df.copy()
+
+
+def series(
+    client: CogniteClient,
+    start: CogniteTimeT,
+    end: CogniteTimeT,
+    aggregate: Optional[str] = None,
+    granularity: Optional[str] = None,
+    include_outside_points: bool = None,
+    limit: Optional[int] = None,
+    query_by: Literal["id", "external_id"] = "id",
+    ignore_unknown_ids: bool = True,
+    ffill: bool = False,
+    fillna: Any = None,
+) -> RetrievalFuncT:
+    """
+    Returns a dict with series of data points.
+
+    This is a very thin wrapper around `CogniteClient.datapoints.retrieve`, with
+    a few tweaks:
+     * if `ignore_unknown_ids=True` is used, the response will always contain
+       all tags that have been requested, even if these tags don't exists in CDF
+     * this function takes `aggregate` (singular) instead of `aggregates`
+       (plural), and expects a string (no support for a list, use multiple `Tag`
+       instances for that).
+     * additional argument: `ffill` will forward-fill the data is set to `True`
+     * additional argument: `fillna` will replace `np.nan` with specified value
+       unless set to `None` (which is the default)
+    """
+
+    def _retrieve(tags: Set[str]) -> TagValueStoreResultT:
+        if tags:
+            df = retrieve_datapoints_df(
+                client,
+                **{query_by: list(tags)},  # type: ignore
+                start=start,
+                end=end,
+                aggregates=[aggregate] if aggregate else None,
+                granularity=granularity,
+                include_outside_points=include_outside_points,
+                ignore_unknown_ids=ignore_unknown_ids,
+                limit=limit,
+            )
+        else:
+            df = pd.DataFrame({})
+        if ffill:
+            df = df.ffill()
+        if fillna is not None:
```

Comment:
He he, and I originally had `undefined = object()` on the module, and `if fillna is not undefined` here so that people could do `fillna` with `None`. Needn't have bothered... 😁